### PR TITLE
[Community pipeline] SDXL Differential Diffusion Img2Img Pipeline

### LINKED ISSUE #7038: Differential Diffusion: Giving Each Pixel Its Strength
### Model/Pipeline/Scheduler description

Hello,
I would like to suggest merging my paper: Differential Diffusion: Giving Each Pixel Its Strength.
The paper allows a user to edit a picture by a change map that describes how much each region should change.
The editing process is typically guided by textual instructions, although it can also be applied without guidance.
We support both continuous and discrete editing.
Our framework is training and fine tuning free! And has negligible penalty of the inference time.
Our implementation is diffusers-based.
 We already tested it on 4 different diffusion models (Kadinsky, DeepFloyd IF, SD, SD XL).
We are confident that the framework can also be ported to other diffusion models, such as SD Turbo, Stable Cascade, and amused.
I notice that you usually stick to white==change convention, which is opposite to the convention we used in the paper.
The paper can be thought of as a generalization to some of the existing techniques.
A black map is just regular txt2img ("0"),
A map of one color (which isn't black) can be thought as img2img,
A map of two colors which one color is white can be thought as inpaint.
And the rest? It's completely new!
In the paper, we suggest some further applications such as soft inpainting and strength visualization.

### Open source status

- [X] The model implementation is available.

### Provide useful links for the implementation

Site:
https://differential-diffusion.github.io/
Paper:
https://differential-diffusion.github.io/paper.pdf
Repo:
https://github.com/exx8/differential-diffusion

### PR DESCRIPTION
**Pull request by asomoza**
# What does this PR do?

Adds [Differential Diffusion](https://github.com/exx8/differential-diffusion) as a community pipeline as discussed [here](https://github.com/huggingface/diffusers/pull/7442#issuecomment-2027753744), this is the SDXL img2img version.

More information and original PR [here](https://github.com/huggingface/diffusers/pull/7442). I'll add the documentation to the README after the initials reviews.

Once this one is merged I plan to do also:

- [X] SDXL Differential Diffusion Img2Img Pipeline
- [ ] SDXL Differential Diffusion Img2Img Controlnet Adapter Pipeline
- [ ] SDXL Differential Diffusion Inpaint Pipeline
- [ ] SDXL Differential Diffusion Inpaint Controlnet Adapter Pipeline

Note: It works with turbo and lighting variations, but IMO the only half good one is the lighting 8 steps, with less steps the model can't properly do the inpaint and even with 8, the inpainting isn't that good.

Fixes #7038 

Paper: https://differential-diffusion.github.io/paper.pdf
Project site: https://differential-diffusion.github.io/
Code: https://github.com/exx8/differential-diffusion

### Example usage:

```python
import torch
from torchvision import transforms

from diffusers import DPMSolverMultistepScheduler
from diffusers.utils import load_image
from examples.community.pipeline_stable_diffusion_xl_differential_img2img import (
    StableDiffusionXLDifferentialImg2ImgPipeline,
)


pipeline = StableDiffusionXLDifferentialImg2ImgPipeline.from_pretrained(
    "SG161222/RealVisXL_V4.0", torch_dtype=torch.float16, variant="fp16"
).to("cuda")
pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config, use_karras_sigmas=True)


def preprocess_image(image):
    image = image.convert("RGB")
    image = transforms.CenterCrop((image.size[1] // 64 * 64, image.size[0] // 64 * 64))(image)
    image = transforms.ToTensor()(image)
    image = image * 2 - 1
    image = image.unsqueeze(0).to("cuda")
    return image


def preprocess_map(map):
    map = map.convert("L")
    map = transforms.CenterCrop((map.size[1] // 64 * 64, map.size[0] // 64 * 64))(map)
    map = transforms.ToTensor()(map)
    map = map.to("cuda")
    return map


image = preprocess_image(
    load_image(
        "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/20240329211129_4024911930.png?download=true"
    )
)

mask = preprocess_map(
    load_image(
        "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/gradient_mask.png?download=true"
    )
)

prompt = "a green pear"
negative_prompt = "blurry"

image = pipeline(
    prompt=prompt,
    negative_prompt=negative_prompt,
    guidance_scale=7.5,
    num_inference_steps=25,
    original_image=image,
    image=image,
    strength=1.0,
    map=mask,
).images[0]

image.save("result.png")
```

### Examples
|Realvis|Juggernaut|
|---|---|
|differential_20240401151708_2861674568|differential_20240401151545_2067481981|
|Lighting 8 steps|IP Adapters|
|differential_20240401145538_3135922217|differential_20240401170302_43613221|

## Who can review?
 @yiyixuxu @sayakpaul @exx8 @a-r-r-o-w

### CODE REVIEW

**Comment by yiyixuxu** [comment]
looks good! will merge whenever you're ready!

**Comment by xinru-liu** [comment]
Hi, I would really like to test the differential img2img pipeline, has this been published yet?

**Comment by yiyixuxu** [comment]
@xinru-liu 
it is merged you can check it out here https://github.com/huggingface/diffusers/tree/main/examples/community#differential-diffusion
we also published a blog about it https://huggingface.co/blog/OzzyGT/outpainting-differential-diffusion

**Comment by xinru-liu** [comment]
@yiyixuxu 
Thank you for getting back, I just tested, it shows 
"404 Client Error: Not Found for url: https://raw.githubusercontent.com/huggingface/diffusers/v0.27.2/examples/community/pipeline_stable_diffusion_xl_differential_img2img.py"
I think the name is accurate, not sure why it failed to access the code.

**Comment by asomoza** [comment]
that's because you're trying to access it in the v0.27.2 release, this was merged after the last release so until the next one you have to install `diffusers` from the git repository:

`pip install git+https://github.com/huggingface/diffusers.git`

Also you can just download it and use it directly:

https://github.com/huggingface/diffusers/blob/main/examples/community/pipeline_stable_diffusion_xl_differential_img2img.py

**Comment by xinru-liu** [comment]
> that's because you're trying to access it in the v0.27.2 release, this was merged after the last release so until the next one you have to install `diffusers` from the git repository:
> 
> `pip install git+https://github.com/huggingface/diffusers.git`
> 
> Also you can just download it and use it directly:
> 
> https://github.com/huggingface/diffusers/blob/main/examples/community/pipeline_stable_diffusion_xl_differential_img2img.py

Thank you @asomoza ! I've tested the solution, it works really well!
